### PR TITLE
Handle empiar starfiles as annotation data

### PR DESCRIPTION
**Pull request by sherwoodf**
Ticket: https://app.clickup.com/t/869a4a328

Adds handling of star files from empiar submission as annotation data. This creates all relevant objects in the same manner as the existing empiar ingest, except for 1 thing: the attribute on the annotation data objects which label which part of the annotation data file correspond to the file is missing. I'm not aware of us using this for anything yet, so i think this passes the requirements of the ticket. I can either do a quick solution to this (which involves generating curation yamls to run curation on the object) but would rather solve this in a large ticket to handle fields that aren't covered by our own fields in general.

Note: does not handle annotation data in ro-crate metadata, only in file lists. I have cut another ticket: https://app.clickup.com/t/869a685dn in which i want to test for & implement the generic case of images being named / identified in both a file list and an ro-crate at the same time. This isn't what the empair or biostudies ingest currently does, but we need to have clearly tested & defined behaviour for it.

### CODE REVIEW

File: ro-crate-ingest/ro_crate_ingest/empiar_to_ro_crate/entity_conversion/file_list.py [hunk 1]
```diff
@@ -132,36 +137,72 @@ def find_matching_imageset_and_images(
...
+            "type": bia_type,
+            "label": label,
+            "source_image_label": source_image_label,
+            "associated_annotation_method": associated_annotation_method,
+            "associated_protocol": associated_protocol,
         }
     )
 
 
 def split_dataframes_by_dataset(file_df: pd.DataFrame):
```
Comment (matthewh-ebi) [inline ro-crate-ingest/ro_crate_ingest/empiar_to_ro_crate/entity_conversion/file_list.py:194]: docstring/return type would help understand what this function's doing

File: ro-crate-ingest/ro_crate_ingest/ro_crate_to_api/api_conversion.py [hunk 2]
```diff
@@ -99,16 +99,16 @@ def convert_ro_crate_to_bia_api(
         get_settings().parallelisation_max_workers,
     )
 
-    if not image_raw_dataframe["image_id"].isna().all():
+    if not image_raw_dataframe["result_data_id"].isna().all():
```
Comment (matthewh-ebi) [inline ro-crate-ingest/ro_crate_ingest/ro_crate_to_api/api_conversion.py:102]: Comment to help understanding of this row would be good - without pandas background knowledge it's hard to parse

File: ro-crate-ingest/ro_crate_ingest/ro_crate_to_api/entity_conversion/file_reference.py [hunk 3]
```diff
@@ -150,23 +151,41 @@ def select_size_in_bytes(row: dict) -> int:
     raise ValueError(f"No dataset found for file: {row["path"]}")
 
 
-def select_image_id(row: dict, image_extensions: list[str]) -> Optional[str]:
+def select_result_data_id_and_type(row: dict, image_extensions: list[str]) -> Optional[str]:
```
Comment (matthewh-ebi) [inline ro-crate-ingest/ro_crate_ingest/ro_crate_to_api/entity_conversion/file_reference.py:154]: Return type wrong here? Looks to be returning a tuple (either of which is None?)

File: ro-crate-ingest/ro_crate_ingest/ro_crate_to_api/entity_conversion/file_reference.py [hunk 4]
```diff
@@ -150,23 +151,41 @@ def select_size_in_bytes(row: dict) -> int:
...
+            bia_type = "http://bia/Image"
 
-    standardised_file_extension = get_standardised_extension(row["path"])
-    if standardised_file_extension in image_extensions:
-        return row["path"]
+    return id, bia_type
 
 
-def select_image_label(row: dict) -> Optional[str]:
+def select_result_data_label(row: dict) -> Optional[str]:
```
Comment (matthewh-ebi) [inline ro-crate-ingest/ro_crate_ingest/ro_crate_to_api/entity_conversion/file_reference.py:188]: Definitely could use docstring here, we use different types of ids in different bits of code and it would help to understand which we're dealing with here

File: ro-crate-ingest/ro_crate_ingest/ro_crate_to_api/entity_conversion/result_data_and_dependency_creation.py [hunk 5]
```diff
@@ -283,3 +231,52 @@ def create_uuid_list(
...
+
+    persist(
+        accession_id,
+        APIModels.AnnotationData,
+        [APIModels.AnnotationData(**model_dict)],
+        persistence_mode,
+    )
+
+
+def caluclate_dependency_height(df):
```
Comment (matthewh-ebi) [inline ro-crate-ingest/ro_crate_ingest/ro_crate_to_api/entity_conversion/result_data_and_dependency_creation.py:264]: caluclate -> calculate, more critically some explanation of what we're doing here (graph traversal things?)

**Review by matthewh-ebi — CHANGES_REQUESTED** [review]
Looks great overall, the requested changes are all readability related.

I don't want to hold up this PR, but for the complex logic we’ll be heavily dependent on in future, we might want some unit type tests to cover edge cases and failure modes. File list processing (particularly matching image / annotation labels for on), and probably the dependency graph processing code as well, that looks gnarly and full of pit traps.

File: ro-crate-ingest/ro_crate_ingest/empiar_to_ro_crate/entity_conversion/file_list.py [hunk 6]
```diff
@@ -132,36 +137,72 @@ def find_matching_imageset_and_images(
...
+                pattern = annotation_data.get("file_pattern", None)
+                if pattern:
+                    result = parse.parse(f"data/{pattern}", path)
+                    if result is not None:
+                        label = annotation_data["label"]
+                        bia_type = "http://bia/AnnotationData"
+                        source_image_label = annotation_data.get(
+                            "input_image_label", None
+                        )
+                        if is_list_of_dicts(source_image_label):
```
[the remark this answers is not in the document]
Comment (sherwoodf) [reply ro-crate-ingest/ro_crate_ingest/empiar_to_ro_crate/entity_conversion/file_list.py:167]: I've realised this is actually where i've changed the case the way the yaml should be formatted, so i can simply the logic. It's actually always yaml objects.
